packages/gas-fee-controller/src/GasFeeController.ts

```diff
@@ -319,48 +317,30 @@ export class GasFeeController extends BaseControllerV2<
     this.pollTokens = new Set();
     this.getCurrentNetworkEIP1559Compatibility =
       getCurrentNetworkEIP1559Compatibility;
-
     this.getCurrentNetworkLegacyGasAPICompatibility =
       getCurrentNetworkLegacyGasAPICompatibility;
-
     this.getCurrentAccountEIP1559Compatibility =
       getCurrentAccountEIP1559Compatibility;
+    this.#getProvider = getProvider;
     this.EIP1559APIEndpoint = EIP1559APIEndpoint;
     this.legacyAPIEndpoint = legacyAPIEndpoint;
     this.clientId = clientId;
+
+    this.ethQuery = new EthQuery(this.#getProvider());
+
     if (onNetworkStateChange && getChainId) {
-      const initialProvider = getProvider();
-      this.ethQuery = new EthQuery(initialProvider);
       this.currentChainId = getChainId();
-      onNetworkStateChange(async () => {
-        const newProvider = getProvider();
-        const newChainId = getChainId();
-        this.ethQuery = new EthQuery(newProvider);
-        if (this.currentChainId !== newChainId) {
-          this.currentChainId = newChainId;
-          await this.resetPolling();
-        }
+      onNetworkStateChange(async (networkControllerState) => {
+        await this.#onNetworkControllerStateChange(networkControllerState);
       });
     } else {
-      const providerConfig = this.messagingSystem.call(
-        'NetworkController:getProviderConfig',
-      );
-      this.currentChainId = providerConfig.chainId;
-      this.ethQuery = this.messagingSystem.call(
-        'NetworkController:getEthQuery',
-      );
-
+      this.currentChainId = this.messagingSystem.call(
+        'NetworkController:getState',
+      ).providerConfig.chainId;
       this.messagingSystem.subscribe(
-        'NetworkController:providerConfigChange',
-        async (newProviderConfig) => {
-          this.ethQuery = this.messagingSystem.call(
-            'NetworkController:getEthQuery',
-          );
-
-          if (this.currentChainId !== newProviderConfig.chainId) {
-            this.currentChainId = newProviderConfig.chainId;
-            await this.resetPolling();
-          }
+        'NetworkController:stateChange',
```

Comment:
Note that we have a similar pattern of using `onNetworkStateChange` or subscribing to an event through the messenger in TokenListController, so I made them match each other.